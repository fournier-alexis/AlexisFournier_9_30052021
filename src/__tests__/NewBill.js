import NewBillUI from "../views/NewBillUI.js";
import { screen, fireEvent } from "@testing-library/dom";
import NewBill from "../containers/NewBill";
import { localStorageMock } from "../__mocks__/localStorage";
import { ROUTES } from "../constants/routes";
import userEvent from "@testing-library/user-event";
import firebase from "../__mocks__/firebase";
import DashboardUI from "../views/DashboardUI";
import BillsUI from "../views/BillsUI";

jest.mock("../app/Firestore");

describe("Given I am connected as an employee", () => {
  let onNavigate;
  let newBill;
  beforeEach(async () => {
    $.fn.modal = jest.fn();
    document.body.innerHTML = NewBillUI();
    onNavigate = (pathname) => {
      document.body.innerHTML = ROUTES({ pathname });
    };
    Object.defineProperty(window, "localStorage", { value: localStorageMock });
    window.localStorage.setItem(
      "user",
      JSON.stringify({
        type: "Employee",
        email: "test@test.test",
      })
    );
    newBill = new NewBill({
      document,
      onNavigate,
      firestore: undefined,
      localStorage: window.localStorage,
    });
  });

  describe("When I am on NewBill Page", () => {
    test("Then i should see the complete form ", () => {
      expect(screen.getByTestId("form-new-bill")).toMatchSnapshot();
    });
    test("I should have all choice in the type select box", () => {
      expect(screen.getAllByRole("option").length).toEqual(7);
    });

    describe("When I chose a file as justify", () => {
      test("I should see the file selected", () => {
        const handleChangeFile = jest.fn(newBill.handleChangeFile);

        const file = screen.getByTestId("file");
        file.addEventListener("change", handleChangeFile);

        fireEvent.change(file, {
          target: {
            files: [new File(["test.jpg"], "test.jpg", { type: "image/jpeg" })],
          },
        });

        expect(handleChangeFile).toHaveBeenCalledTimes(1);
        expect(file.files[0].name).toBe("test.jpg");
      });
      test("I should have an error", () => {
        jest.spyOn(window, "alert").mockImplementation(() => {});
        const handleChangeFile = jest.fn(newBill.handleChangeFile);

        const file = screen.getByTestId("file");
        file.addEventListener("change", handleChangeFile);

        fireEvent.change(file, {
          target: {
            files: [new File(["test.txt"], "test.txt", { type: "text/txt" })],
          },
        });

        expect(window.alert).toBeCalled();
      });
    });
    describe("When i send the form", () => {
      test("I should be redirect to the bill page", () => {
        newBill = new NewBill({
          document,
          onNavigate,
          firestore: undefined,
          localStorage,
        });
        screen.getByTestId("expense-type").value = "Transports";
        screen.getByTestId("expense-name").value = "test bill";
        screen.getByTestId("amount").value = "10";
        screen.getByTestId("datepicker").value = "2021-06-07";
        screen.getByTestId("vat").value = "";
        screen.getByTestId("pct").value = "20";
        screen.getByTestId("commentary").value = "test commentary";
        const handleSubmit = jest.fn(newBill.handleSubmit);
        const form = screen.getByTestId("form-new-bill");
        form.addEventListener("click", handleSubmit);
        userEvent.click(form);
        expect(handleSubmit).toHaveBeenCalled();
        expect(screen.getByText("Mes notes de frais")).toBeTruthy();
      });
    });

    // test d'intÃ©gration Post
    describe("When I create a new bill", () => {
      const bill = {
        email: "test@test.test",
        type: "Transports",
        name: "test",
        amount: 0,
        date: Date.now(),
        vat: 0,
        pct: 20,
        commentary: "",
        fileUrl: "",
        fileName: "",
        status: "pending",
      };

      test("create a new bill POST", async () => {
        const postSpy = jest.spyOn(firebase, "post");
        const bills = await firebase.post(bill);
        expect(postSpy).toHaveBeenCalledTimes(1);
        expect(bills.data.length).toEqual(5);
      });

      test("fetches bills from an API and fails with 404 message error", async () => {
        firebase.post.mockImplementationOnce(() =>
          Promise.reject(new Error("Erreur 404"))
        );
        const html = BillsUI({ error: "Erreur 404" });
        document.body.innerHTML = html;
        const message = await screen.getByText(/Erreur 404/);
        expect(message).toBeTruthy();
      });

      test("fetches messages from an API and fails with 500 message error", async () => {
        firebase.post.mockImplementationOnce(() =>
          Promise.reject(new Error("Erreur 500"))
        );
        const html = BillsUI({ error: "Erreur 500" });
        document.body.innerHTML = html;
        const message = await screen.getByText(/Erreur 500/);
        expect(message).toBeTruthy();
      });
    });
  });
});
