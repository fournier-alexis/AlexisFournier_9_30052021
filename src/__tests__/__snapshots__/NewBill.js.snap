// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given I am connected as an employee When I am on NewBill Page Then i should see the complete form  1`] = `
<form
  data-testid="form-new-bill"
>
  
            
  <div
    class="row"
  >
    
                
    <div
      class="col-md-6"
    >
      
                  
      <div
        class="col-half"
      >
        
                    
        <label
          class="bold-label"
          for="expense-type"
        >
          Type de dépense
        </label>
        
                      
        <select
          class="form-control blue-border"
          data-testid="expense-type"
          required=""
        >
          
                        
          <option>
            Transports
          </option>
          
                        
          <option>
            Restaurants et bars
          </option>
          
                        
          <option>
            Hôtel et logement
          </option>
          
                        
          <option>
            Services en ligne
          </option>
          
                        
          <option>
            IT et électronique
          </option>
          
                        
          <option>
            Equipement et matériel
          </option>
          
                        
          <option>
            Fournitures de bureau
          </option>
          
                      
        </select>
        
                  
      </div>
      
                  
      <div
        class="col-half"
      >
        
                    
        <label
          class="bold-label"
          for="expense-name"
        >
          Nom de la dépense
        </label>
        
                    
        <input
          class="form-control blue-border"
          data-testid="expense-name"
          placeholder="Vol Paris Londres"
          type="text"
        />
        
                  
      </div>
      
                  
      <div
        class="col-half"
      >
        
                    
        <label
          class="bold-label"
          for="datepicker"
        >
          Date
        </label>
        
                    
        <input
          class="form-control blue-border"
          data-testid="datepicker"
          required=""
          type="date"
        />
        
                  
      </div>
      
                  
      <div
        class="col-half"
      >
        
                    
        <label
          class="bold-label"
          for="amount"
        >
          Montant TTC 
        </label>
        
                    
        <input
          class="form-control blue-border input-icon input-icon-right"
          data-testid="amount"
          placeholder="348"
          required=""
          type="number"
        />
        
                  
      </div>
      
                  
      <div
        class="col-half-row"
      >
        
                    
        <div
          class="flex-col"
        >
           
                      
          <label
            class="bold-label"
            for="vat"
          >
            TVA
          </label>
          
                      
          <input
            class="form-control blue-border"
            data-testid="vat"
            placeholder="70"
            type="number"
          />
          
                    
        </div>
        
                    
        <div
          class="flex-col"
        >
          
                      
          <label
            class="white-text"
            for="pct"
          >
            %
          </label>
          
                      
          <input
            class="form-control blue-border"
            data-testid="pct"
            placeholder="20"
            required=""
            type="number"
          />
          
                    
        </div>
        
                  
      </div>
      
                
    </div>
    
                
    <div
      class="col-md-6"
    >
      
                  
      <div
        class="col-half"
      >
        
                    
        <label
          class="bold-label"
          for="commentary"
        >
          Commentaire
        </label>
        
                    
        <textarea
          class="form-control blue-border"
          data-testid="commentary"
          rows="3"
        />
        
                  
      </div>
      
                  
      <div
        class="col-half"
      >
        
                    
        <label
          class="bold-label"
          for="file"
        >
          Justificatif
        </label>
        
                    
        <input
          class="form-control blue-border"
          data-testid="file"
          required=""
          type="file"
        />
        
                  
      </div>
      
                
    </div>
    
            
  </div>
  
            
  <div
    class="row"
  >
    
              
    <div
      class="col-md-6"
    >
      
                
      <div
        class="col-half"
      >
        
                  
        <button
          class="btn btn-primary"
          id="btn-send-bill"
          type="submit"
        >
          Envoyer
        </button>
        
                
      </div>
      
              
    </div>
    
            
  </div>
  
          
</form>
`;
